### ISSUE TITLE
QA: Mount Reverse with ability to choose between encrypted and plaintext names?

### ISSUE THREAD

**Issue by jdr85** [rfjakob/gocryptfs#561]
Hi,

I would like to reverse mount my unencrypted directory to an encrypted one with the ability to choose whether I want the filenames encrypted or not. The reason for this is an application which crashes when the path is too long so I need to disable filename encryption in this case.

Can I just do the following:
- gocryptfs -reverse -init /unencrypted_dir  # Creates .gocryptfs.reverse.conf with encrypted names 
- rename this to .gocryptfs.reverse_encryptednames.conf
- copy .gocryptfs.reverse_encryptednames.conf to .gocryptfs.reverse_plaintextnames.conf
- In .gocryptfs.reverse_plaintextnames.conf only change 
   "FeatureFlags": [
                "GCMIV128",
                "HKDF",
                "DirIV",
                "EMENames",
                "LongNames",
                "Raw64",
                "AESSIV"
        ]
to
    "FeatureFlags": [
                "GCMIV128",
                "HKDF",
                "PlaintextNames",
                "AESSIV"
        ]
leaving everything else (especially EncryptedKey and ScryptObject) the same
- Choose before each mount which .gocryptfs.reverse.conf (plaintext or encrypted name) I want to use and rename the respective file to .gocryptfs.reverse.conf.

Would that work?
Would /gocryptreverse/somefile.txt (without filename encryption) and /gocryptreverse/encrypted_name_of_somefile.txt (with filename encryption) be the same files?

**Comment by rfjakob** [comment]
Hi, you can use `-config` to chose which config file to use, this saves the rename step.

But the two files you mention,

* /gocryptreverse/somefile.txt (without filename encryption)
* /gocryptreverse/encrypted_name_of_somefile.txt (with filename encryption)

would have different encrypted content because the file path is used as the initialization vector.

**Comment by rfjakob** [comment]
PS: Code is here, and n.Path() is not encrypted in the first case and encrypted in the 2nd case

https://github.com/rfjakob/gocryptfs/blob/5cb1e55714aa92a848c0fb5fc3fa7b91625210fe/internal/fusefrontend_reverse/node.go#L147-L148

**Comment by jdr85** [comment]
Thanks @rfjakob 
Unfortunately I don't speak Go but what you write makes sense. :)

I already tried the -config directive, but found that in this case there is no gocryptfs.conf in the reverse encrypted directory. Therefore I would need to mount my encrypted backup with the master key:
`root@jdr-z83:~/test# gocryptfs -reverse -config unencrypted/.gocryptfs.reverse_fnplaintext.conf unencrypted encrypted`
`root@jdr-z83:~/test/encrypted# ls -Al`
`total 12`
`-r-------- 1 root root 447 Apr  7 20:08 .gocryptfs.reverse_fnencrypted.conf`
`-r-------- 1 root root 416 Apr  7 20:09 .gocryptfs.reverse_fnplaintext.conf`
`-rw-r--r-- 1 root root  55 Apr  7 20:07 test.txt`

**Comment by rfjakob** [comment]
Oh, that's true, but note that the gocryptfs.conf that appears in the encrypted dir is just a copy of .gocryptfs.reverse.conf.

So you could also mount in forward mode like this:

    gocryptfs -config unencrypted/.gocryptfs.reverse_fnplaintext.conf encrypted unencrypted

**Comment by jdr85** [comment]
Sorry for my late response. 
I like my encrypted backups to be self-contained and I am afraid to forget manually addding the .gocryptfs.reverse.whatever.conf to my backup when I switch to another backup target (which sometimes happens).

Therefore I decided to not use the -config directive but instead copy the conf file I need (with or without filename encryption) to .gocryptfs.reverse.conf before mounting. To be more precise, the unencrypted folders are synced between 2 systems. On one system I need to reverse mount it with filename encryption, on the other system I need to reverse mount it with plaintext names. I just copied the ...fnencrypted.conf on system A and the ...fnplaintext on system B to .gocryptfs.reverse.conf and excluded this file from sync. This works perfectly. :)
